java/com.oracle.mxtool.compilerserver/src/com/oracle/mxtool/compilerserver/JavacDaemon.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.mxtool.compilerserver;
+
+import java.io.BufferedReader;
+import java.io.DataOutputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Method;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class JavacDaemon implements Runnable {
+
+    private static void logf(String commandLine, Object... args) {
+        if (verbose) {
+            System.err.printf(commandLine, args);
+        }
+    }
+
+    private static boolean verbose = false;
+    private static boolean running;
+    private static ThreadPoolExecutor threadPool;
+
+    private Socket connectionSocket;
+    private Object compiler;
+    private Method compile;
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Usage: port");
+            System.exit(1);
+        }
+
+        // create socket
+        int port = Integer.parseInt(args[0]);
+        ServerSocket serverSocket = new ServerSocket();
+        serverSocket.setReuseAddress(true);
+        serverSocket.bind(new InetSocketAddress(port));
+        logf("Started server on port %d\n", port);
+
+        final Class<?> javacMainClass = Class.forName("com.sun.tools.javac.Main");
+        final Method compileMethod = javacMainClass.getMethod("compile", new Class<?>[]{(new String[]{}).getClass()});
+
+        int threadCount = Runtime.getRuntime().availableProcessors();
+        threadPool = new ThreadPoolExecutor(threadCount, threadCount, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(), runnable -> new Thread(runnable));
+
+        running = true;
+        while (running) {
+            Socket connectionSocket = serverSocket.accept();
+            threadPool.submit(new JavacDaemon(connectionSocket, javacMainClass.newInstance(), compileMethod));
+        }
+        serverSocket.close();
+    }
+
+    public JavacDaemon(Socket connectionSocket, Object compiler, Method compile) {
+        this.connectionSocket = connectionSocket;
+        this.compiler = compiler;
+        this.compile = compile;
+    }
+
+    public void run() {
+        try {
+            BufferedReader input = new BufferedReader(new InputStreamReader(connectionSocket.getInputStream()));
+            DataOutputStream output = new DataOutputStream(connectionSocket.getOutputStream());
+
+            try {
+                String commandLine = input.readLine();
```

Comment:
I hope Java and python will agree on the charset used to serialize/deserialize those strings.